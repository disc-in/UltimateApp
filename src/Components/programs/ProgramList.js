import React, { useState } from 'react';
import { StyleSheet, Text, View, ScrollView, TouchableOpacity } from 'react-native';
import Accordion from 'react-native-collapsible/Accordion';
import { MaterialCommunityIcons } from '@expo/vector-icons';

import theme from '../../styles/theme.style';
import Program from './Program';

const ProgramList = (props) => {
  const { navigation, displayedPrograms, activeProgramId, completeTrainings } = props;

  const activeProgramIndex = displayedPrograms.findIndex((program) => program.id === activeProgramId);
  const [activeSections, setActiveSections] = useState([activeProgramIndex]);

  const setSections = (sections) => {
    setActiveSections(sections);
  };

  const renderHeader = (section, _, isActive) => {
    return <Program program={section} />;
  };

  const renderTraining = (training, index) => {
    const program = displayedPrograms[activeSections];
    const onTrainingPress = () => navigation.navigate('TrainingPage', { training, program });

    const isDone = program
      ? completeTrainings.find(
          (complete) => program.id === complete.program.id && training.id === complete.training.id,
        ) !== undefined
      : false;

    return (
      <TouchableOpacity
        key={training.id.toString()}
        style={[styles.training, isDone ? styles.trainingDone : null]}
        onPress={onTrainingPress}
      >
        <View style={styles.descriptionContainer}>
          <Text style={styles.trainingTitle}>
            {index + 1} - {training.title}
          </Text>
          <Text style={styles.trainingDescription}>{training.description}</Text>
        </View>
        <View style={styles.arrowDoneContainer}>
          {isDone && <MaterialCommunityIcons name="check-circle" color={theme.COLOR_PRIMARY} size={26} />}
          <MaterialCommunityIcons name="chevron-right" color={theme.COLOR_PRIMARY} size={26} />
        </View>
      </TouchableOpacity>
    );
  };

  const renderContent = (section, _, isActive) => {
    return section.trainings.map((training, index) => {
      return renderTraining(training, index);
    });
  };

  return (
    <ScrollView style={styles.programList}>
      <Accordion
        activeSections={activeSections}
        sections={displayedPrograms}
        touchableComponent={TouchableOpacity}
        expandMultiple={false}
        renderHeader={renderHeader}
        renderContent={renderContent}
        duration={400}
        onChange={setSections}
        containerStyle={styles.accordionContainerStyle}
      />
    </ScrollView>
  );
};

export default ProgramList;

const styles = StyleSheet.create({
  programList: {
    height: '100%',
  },
  accordionContainerStyle: {
    marginBottom: 10,
  },
  training: {
    padding: 10,
    paddingLeft: 30,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
    backgroundColor: theme.COLOR_PRIMARY_LIGHT,
    borderBottomColor: theme.COLOR_SECONDARY_LIGHT,
    borderBottomWidth: 1,
  },
  trainingDone: {
    backgroundColor: '#f5fbfc',
  },
  descriptionContainer: {
    flexBasis: '80%',
  },
  trainingTitle: {
    fontSize: theme.FONT_SIZE_MEDIUM,
    fontWeight: 'bold',
    color: theme.COLOR_PRIMARY,
  },
  trainingDescription: {
    fontSize: theme.FONT_SIZE_SMALL,
    color: theme.COLOR_PRIMARY,
  },
  arrowDoneContainer: {
    flexBasis: '20%',
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'flex-end',
  },
});
