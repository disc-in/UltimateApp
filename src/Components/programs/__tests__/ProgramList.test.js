import React from 'react';
import { render } from '@testing-library/react-native';
import { Provider } from 'react-redux';

import { createProgram, createTraining } from '../../../Fixtures/TestFixtures';
import store from '../../../Store/testStore';

import ProgramList from '../ProgramList';

describe('<ProgramList />', () => {
  it('renders correctly', () => {
    const navigation = { navigate: jest.fn() };

    const program1 = createProgram({ title: 'Program 1' });
    const firstTraining = createTraining({ id: 1, title: 'First Training' });
    const secondTraining = createTraining({ id: 2, title: 'Second Training' });
    const program2 = createProgram({ title: 'Program 2', trainings: [firstTraining, secondTraining] });
    const programs = [program1, program2];

    const { toJSON } = render(
      <Provider store={store}>
        <ProgramList displayedPrograms={programs} navigation={navigation} completeTrainings={[]} />
      </Provider>,
    );
    expect(toJSON()).toMatchSnapshot();
  });
});
