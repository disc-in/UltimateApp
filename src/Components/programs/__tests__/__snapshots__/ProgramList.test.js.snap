// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProgramList /> renders correctly 1`] = `
<RCTScrollView
  style={
    Object {
      "height": "100%",
    }
  }
>
  <View>
    <View
      style={
        Object {
          "marginBottom": 10,
        }
      }
    >
      <View
        style={Object {}}
      >
        <View
          accessibilityState={
            Object {
              "expanded": false,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Object {
                "backgroundColor": "#fff",
                "elevation": 3,
                "marginTop": 10,
                "padding": 10,
                "shadowOffset": Object {
                  "height": 1,
                  "width": 0,
                },
                "shadowOpacity": 0.3,
                "shadowRadius": 3,
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "flex-start",
                  "backgroundColor": "#fff",
                  "minHeight": 40,
                  "paddingBottom": 10,
                  "paddingHorizontal": 20,
                  "width": "100%",
                }
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "flex": 1,
                    "flexDirection": "row",
                    "justifyContent": "space-between",
                    "width": "100%",
                  }
                }
              >
                <Text
                  style={
                    Object {
                      "color": "#232b2b",
                      "fontSize": 20,
                      "fontWeight": "bold",
                    }
                  }
                >
                  Program 1
                </Text>
                <Text
                  style={
                    Object {
                      "color": "#232b2b",
                      "fontSize": 14,
                      "textAlign": "right",
                    }
                  }
                >
                  0/1
                </Text>
              </View>
              <View
                style={
                  Object {
                    "backgroundColor": "#f1f1f1",
                    "borderRadius": 2,
                    "height": 3,
                    "marginTop": 5,
                    "width": "100%",
                  }
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "bottom": 0,
                        "left": 0,
                        "position": "absolute",
                        "right": 0,
                        "top": 0,
                      },
                      Object {
                        "backgroundColor": "#2a9e91",
                        "borderRadius": 5,
                        "width": "50%",
                      },
                      Object {
                        "width": "0%",
                      },
                    ]
                  }
                />
              </View>
            </View>
          </View>
        </View>
        <View
          collapsable={false}
          pointerEvents="none"
          style={
            Object {
              "height": 0,
              "overflow": "hidden",
            }
          }
        >
          <View
            collapsable={false}
            forwardedRef={[Function]}
            onLayout={[Function]}
            style={Object {}}
          >
            <View
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#fff",
                  "borderBottomColor": "#e5e5e5",
                  "borderBottomWidth": 1,
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                  "opacity": 1,
                  "padding": 10,
                  "paddingLeft": 30,
                }
              }
            >
              <View
                style={
                  Object {
                    "flexBasis": "80%",
                  }
                }
              >
                <Text
                  style={
                    Object {
                      "color": "#232b2b",
                      "fontSize": 16,
                      "fontWeight": "bold",
                    }
                  }
                >
                  1
                   - 
                  First practice
                </Text>
                <Text
                  style={
                    Object {
                      "color": "#232b2b",
                      "fontSize": 14,
                    }
                  }
                >
                  This training session aims at..
                </Text>
              </View>
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "flexBasis": "20%",
                    "flexDirection": "row",
                    "justifyContent": "flex-end",
                  }
                }
              >
                <Text />
              </View>
            </View>
          </View>
        </View>
      </View>
      <View
        style={Object {}}
      >
        <View
          accessibilityState={
            Object {
              "expanded": false,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Object {
                "backgroundColor": "#fff",
                "elevation": 3,
                "marginTop": 10,
                "padding": 10,
                "shadowOffset": Object {
                  "height": 1,
                  "width": 0,
                },
                "shadowOpacity": 0.3,
                "shadowRadius": 3,
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "flex-start",
                  "backgroundColor": "#fff",
                  "minHeight": 40,
                  "paddingBottom": 10,
                  "paddingHorizontal": 20,
                  "width": "100%",
                }
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "flex": 1,
                    "flexDirection": "row",
                    "justifyContent": "space-between",
                    "width": "100%",
                  }
                }
              >
                <Text
                  style={
                    Object {
                      "color": "#232b2b",
                      "fontSize": 20,
                      "fontWeight": "bold",
                    }
                  }
                >
                  Program 2
                </Text>
                <Text
                  style={
                    Object {
                      "color": "#232b2b",
                      "fontSize": 14,
                      "textAlign": "right",
                    }
                  }
                >
                  0/2
                </Text>
              </View>
              <View
                style={
                  Object {
                    "backgroundColor": "#f1f1f1",
                    "borderRadius": 2,
                    "height": 3,
                    "marginTop": 5,
                    "width": "100%",
                  }
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "bottom": 0,
                        "left": 0,
                        "position": "absolute",
                        "right": 0,
                        "top": 0,
                      },
                      Object {
                        "backgroundColor": "#2a9e91",
                        "borderRadius": 5,
                        "width": "50%",
                      },
                      Object {
                        "width": "0%",
                      },
                    ]
                  }
                />
              </View>
            </View>
          </View>
        </View>
        <View
          collapsable={false}
          pointerEvents="none"
          style={
            Object {
              "height": 0,
              "overflow": "hidden",
            }
          }
        >
          <View
            collapsable={false}
            forwardedRef={[Function]}
            onLayout={[Function]}
            style={Object {}}
          >
            <View
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#fff",
                  "borderBottomColor": "#e5e5e5",
                  "borderBottomWidth": 1,
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                  "opacity": 1,
                  "padding": 10,
                  "paddingLeft": 30,
                }
              }
            >
              <View
                style={
                  Object {
                    "flexBasis": "80%",
                  }
                }
              >
                <Text
                  style={
                    Object {
                      "color": "#232b2b",
                      "fontSize": 16,
                      "fontWeight": "bold",
                    }
                  }
                >
                  1
                   - 
                  First Training
                </Text>
                <Text
                  style={
                    Object {
                      "color": "#232b2b",
                      "fontSize": 14,
                    }
                  }
                >
                  This training session aims at..
                </Text>
              </View>
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "flexBasis": "20%",
                    "flexDirection": "row",
                    "justifyContent": "flex-end",
                  }
                }
              >
                <Text />
              </View>
            </View>
            <View
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#fff",
                  "borderBottomColor": "#e5e5e5",
                  "borderBottomWidth": 1,
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                  "opacity": 1,
                  "padding": 10,
                  "paddingLeft": 30,
                }
              }
            >
              <View
                style={
                  Object {
                    "flexBasis": "80%",
                  }
                }
              >
                <Text
                  style={
                    Object {
                      "color": "#232b2b",
                      "fontSize": 16,
                      "fontWeight": "bold",
                    }
                  }
                >
                  2
                   - 
                  Second Training
                </Text>
                <Text
                  style={
                    Object {
                      "color": "#232b2b",
                      "fontSize": 14,
                    }
                  }
                >
                  This training session aims at..
                </Text>
              </View>
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "flexBasis": "20%",
                    "flexDirection": "row",
                    "justifyContent": "flex-end",
                  }
                }
              >
                <Text />
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
